rfcs/20180824-tf-print-v2.md

```diff
@@ -0,0 +1,549 @@
+printing and logging as some of their most important tools. Unfortunately,
+setting up printing of Tensors when building TensorFlow graphs doesn't align
+with the natural usage of print primitives most programmers are used to. This
+has led to various bugs, questions on online forums, and entire blog posts
+explaining tf.Print (e.g.
+[blog post 1](https://towardsdatascience.com/using-tf-print-in-tensorflow-aa26e1cff11e),
+[blog post 2](http://www.heyuhang.com/blog/2018/01/31/the-secret-of-tf-dot-print-in-tensorflow/),
+[Github question](https://github.com/tensorflow/tensorflow/issues/1988),
+[Quora question](https://www.quora.com/How-does-the-tf-Print-statement-work-for-TensorFlow),
+[Stack Overflow question](https://stackoverflow.com/questions/33633370/how-to-print-the-value-of-a-tensor-object-in-tensorflow),
+...). Much of the confusion comes from the existing `tf.Print` operator being an
+identity operator with a side effect of printing. This is a very non-standard
+API for printing, and the operator graph must be carefully set up to ensure the
+operator gets executed. This is at odds with printing/logging users are used to
+from elsewhere, where a print method produces no outputs and immediately prints
+the desired value.
+
+Eager execution promises a more interactive, easier to debug experience that
+works with Python's `print` method. However, even with eager execution enabled,
+there is need to wrap parts of execution and optimized graph functions
+(currently `tf.contrib.eager.defun`). The Python `print` method placed in these
+graph function will run at graph construction time, not graph execution time,
+creating confusion.
+
+This doc proposes a new `tf.print` TensorFlow printing approach that is very
+similar to the standard python `print` API whether or not executing eagerly. It
+also provides long-requested functionality for both eager and session-based
+execution, such as a more meaningful Tensor summarization, support for printing
+nested data structures that contain tensors, and controllable logging levels.
+
+## Overview
+
+We introduce two methods to the public Python API: `tf.strings.format` and
+`tf.print`. (We may have to use `tf.strings.Format` or `tf.strings.fmt` instead
+of `tf.strings.format` if the built-in python `format` method causes issues with
+the naming.) These are internally backed by two C++ operators: `StringFormat`
+and `PrintV2`. The method headers & docstrings are as follows:
+
+### Python Public API Methods
+
+````python
+@tf_export("print")
+def print_v2(*inputs, **kwargs):
+  """Print the specified inputs.
+  Returns an operator that prints the specified inputs to a desired
+  output stream or logging level. The inputs may be dense or sparse Tensors,
+  primitive python objects, data structures that contain Tensors, and printable
+  python objects. Printed tensors will recursively show the first and last
+  `summarize` elements of each dimension.
```

Comment:
Is there any way to print the whole elements of a tensor, say, summarize=None or -1 ?